### PR TITLE
Sparse square sum

### PR DESCRIPTION
**Pull request by reminisce**
This PR implements square_sum op for row-sparse matrices. It actually fuses operators `square` and `sum` together to avoid computing dense gradient of the operator `sum` in backward passes. The operator `_square_sum` is registered as an internal operator since it's a temporary solution. After the functionality of fusing operators is finished in the future, this operator may become deprecated.

Supported use cases:

- `_square_sum(rsp, axis=0, keepdims=False/True) -> dns output`
- `_square_sum(rsp, axis=1, keepdims=False) -> dns output`
- `_square_sum(rsp, axis=1, keepdims=True) -> rsp output`

@piiswrong @eric-haibin-lin @madjam @cjolivier01 @anirudh2290 @stefanhenneking

### CODE REVIEW

File: src/operator/tensor/square_sum.cu [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+
+namespace mxnet {
+namespace op {
+
+NNVM_REGISTER_OP(_square_sum)
+.set_attr<FComputeEx>("FComputeEx<gpu>", SquareSumOpForwardEx<gpu>);
+
+NNVM_REGISTER_OP(_backward_square_sum)
+.set_attr<FComputeEx>("FComputeEx<gpu>", SquareSumOpBackwardEx<gpu>);
+
```
Comment (eric-haibin-lin) [inline src/operator/tensor/square_sum.cu]: please remove gpu operators if they're not unit-tested.

**Comment by reminisce** [comment]
GPU unit test cannot be run due to functions missing in `check_numeric_gradient` for sparse tensors on GPU. GPU registration is removed for now and will be added back in the future.

File: src/operator/tensor/square_sum.cu (hunk 1 above)
Comment (reminisce) [inline src/operator/tensor/square_sum.cu]: Done.

File: python/mxnet/test_utils.py [hunk 2]
```diff
@@ -552,7 +552,10 @@ def random_projection(shape):
         assert isinstance(grad_stype_dict, dict), "grad_stype_dict must be a dict"
         for k, v in grad_stype_dict.items():
             if k in args_grad and v in _STORAGE_TYPE_STR_TO_ID and v != 'default':
-                args_grad[k] = mx.nd.cast_storage(args_grad[k], stype=v)
+                # create an uninitialized sparse ndarray for executor
+                # if the symbolic grad is expected to be zero, it should not be initialized at all
+                args_grad[k] = mx.nd.zeros(args_grad[k].shape, args_grad[k].context,
+                                           args_grad[k].dtype, v)
```
Comment (eric-haibin-lin) [inline python/mxnet/test_utils.py]: Is this necessary? Was the previous version using cast_storage not working?

File: src/operator/tensor/square_sum-inl.h [hunk 3]
```diff
@@ -0,0 +1,340 @@
...
+    }
+    return;
+  }
+
+  if (output->storage_type() == kRowSparseStorage) {
+    output->CheckAndAlloc({input.aux_shape(rowsparse::kIdx)});
+  }
+  const TBlob& out_data = output->data();
+  const int64_t nnr = input.storage_shape()[0];
+  const int64_t num_cols = input.storage_shape()[1];
```
Comment (eric-haibin-lin) [inline src/operator/tensor/square_sum-inl.h]: Is it still correct for k-dimensional inputs? Are you assuming 2d inputs here?

File: python/mxnet/test_utils.py (hunk 2 above)
Comment (reminisce) [inline python/mxnet/test_utils.py]: Yes, it is the correct way to do it. The previous version is not correct. The args_grad are outputs in backward FComputeEx functions. They must be CheckAndAlloc inside the FComputeEx functions. The cast_storage would actually create an initialized sparse tensor for the backward which is not right. One corner case illustrating this is that, if the input is uninitialized, the square_sum backward function returns immediately, so the output should be uninitialized as well.

File: src/operator/tensor/square_sum-inl.h (hunk 3 above)
Comment (reminisce) [inline src/operator/tensor/square_sum-inl.h]: I assume the storage_shape for row-sparse storage type is 2D regardless of the dimension of the shape, which is more intuitively understandable.

**Comment by cjolivier01** [comment]
for multiple forward/backward passes, would the input grad arrays (output
of the backward pass) be initialized and still contain data from the
previous pass?

**Comment by reminisce** [comment]
I think so. That's why it's needed to call `FillComputeZerosEx` before `return` when the input is not initialized, which I forgot to do. I will submit another pr to fix this issue for this op. Thanks.

**Comment by cjolivier01** [comment]
I don't know if that's necessary (maybe it is -- I don't know your code as
well as you do). I just was referring to the statement that it was sort of
a requirement that the sparse tensor not be initialized yet, and was
wondering if it was a problem of reshaping on-the-fly or, like you said,
values I the tensor might have garbage values in them and it's not
reconciled. Is there something about the algorithm that would not touch
every active value in the array and leave the garbage value in the result?
My naive expectation is that every stored value would be written to during
the algorithm execution. Just wondering for my own education here.

**Pull request merged**